Open Images Evaluation Tutorial

tylerganter opened the pull request:
## What changes are proposed in this pull request?

A tutorial for evaluating an object detection model on Open Images dataset on a per-image granularity using FiftyOne.

## How is this patch tested? If it is not, in a single sentence please explain why.

This isn't the easiest to test...you can run through the tutorial but it will involve download 36GB of data.

## Release Notes

### Is this a user-facing change?

-   [ ] No. You can skip the rest of this section.
-   [X] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

A new tutorial in the documentation

### What areas of FiftyOne does this PR affect?

-   [ ] `App`: FiftyOne application changes
-   [ ] `Build`: Build and test infrastructure changes
-   [ ] `Core`: Core `fiftyone` Python library changes
-   [X] `Documentation`: FiftyOne documentation changes
-   [ ] `Other`

### Should this PR be mentioned in the release notes? If so, please choose on category:

-   [ ] `breaking-change` - The PR will be mentioned in the "Breaking Changes"
        section
-   [ ] `feature` - A new user-facing feature worth mentioning in the release
        notes
-   [ ] `bug-fix` - A user-facing bug fix worth mentioning in the release notes
-   [x] `documentation` - A user-facing documentation change worth mentioning
        in the release notes

ehofesmann commented on docs/source/tutorials/open_images_eval/error_analysis/inference/tfhub.py:
```diff
@@ -0,0 +1,246 @@
+            model_handle: url to the model, starting with: https://tfhub.dev/
+        """
+        self.model_handle = model_handle
+        self._detector = hub.load(self.model_handle).signatures["default"]
```
After installing multiple tensorflow versions due to various errors, this gave me a permission denied error writing to `/tmp` error from `tensorflow_hub` that finally broke me from trying to generate my own predictions. Not really anything to do here, just venting about tensorflow.

tylerganter replied:
"anyone can do data science"

ehofesmann commented on docs/source/tutorials/open_images_evaluation.ipynb:
```diff
@@ -0,0 +1,444 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pip install bash_kernel\n",
```
I wasn't able to get this to play nice with my existing virtual environment so I had to reinstall a bunch of packages. I'd recommend just having users copy and paste the commands into their console. If you do wanna go with this then probably add instructions of how to use the `Bash` kernel that is created (e.g. "click on the Kernels option in jupyter",...)

Also when I started this notebook, it said that it could not find a kernel called `Bash` because I hadn't installed it yet. Not sure where that is saved in the notebook but the default `Bash` kernel should probably be removed.

tylerganter replied:
I updated this description. Let me know if it's more clear

ehofesmann replied:
The description is good. But see my other comments about maybe removing the `Bash` kernel altogether.

ehofesmann commented on docs/source/tutorials/open_images_evaluation.ipynb:
```diff
@@ -0,0 +1,444 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pip install Pillow tensorflow-hub"
```
Your scripts also need `pandas` that I didn't have installed.

ehofesmann commented on docs/source/tutorials/open_images_evaluation.ipynb:
```diff
@@ -0,0 +1,444 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "aws s3 --no-sign-request sync s3://open-images-dataset/test [target_dir/test]"
```
Just replace `[target_dir/test]` with something concrete like `open-images-fiftyone/test` to make it so you can run it directly.

ehofesmann commented on docs/source/tutorials/open_images_evaluation.ipynb:
```diff
@@ -0,0 +1,444 @@
+    "PREDICTIONS_PATH=/PATH/TO/PREDICTIONS.csv\n",
+    "CLASS_DESCRIPTIONS=/PATH/TO/class-descriptions-boxable.csv\n",
+    "\n",
+    "python open_images_eval/scripts/load_data.py \\\n",
```
Give the command for running this without predictions for people who just want to browse through the dataset.

tylerganter replied:
I wrote a comment in the code block

ehofesmann replied:
You also need to remove `--load_images_with_preds`. Could you just make this an alternate code block to make it easier to copy and paste?

ehofesmann commented on docs/source/tutorials/open_images_evaluation.ipynb:
```diff
@@ -0,0 +1,444 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pip install tensorflow\n",
```
So many tensorflow issues..... I started with v1.12.0 and it gave me a `ModuleNotFoundError: No module named 'tensorflow.python.training.tracking'` when trying to run inference. 

v1.13 gave me cuda issues

v2.1 finally worked for me but then I ran into other issues with `tensorflow_hub` where it was trying to access `/tmp` and couldn't.

It seems like the only outcome of this is to generate `predictions.csv`. I think you should just provide that to be able to run this tutorial quickly and then keep the the inference only for people who really want to run their own model.

Honestly, it seems like you could remove any examples of running inference or using `tensorflow_hub` from the tutorial and just mention that you have a script available if someone would want to do that. I think provding a `predictions.csv` file and describing the format of it is enough.

tylerganter replied:
I changed this to `pip install --upgrade tensorflow` which should resolve version issues. General comment: for the most part in 2020, if you try to use tensorflow v1.X you're going to have a bad time.

What was the issue with `/tmp`?

I also changed the title of the section to `(optional) Generate predictions` and made "**Alternatively**, ..." bold so that it's a little more attention grabbing.

But I think it's fine to leave this section and not have a `predictions.csv` download. The entire purpose of this tutorial is to be able to do this all yourself, in particular, to evaluate your own model. I would like to get the download with 500 images with predictions to be available and probably link that in the top of this tutorial though

ehofesmann replied:
I still feel pretty adamant about providing a `predictions.csv` file.

1) Sure, some people go through tutorials to be able to replicate exactly what is in them for their purposes, but there is also a good portion of people that go through tutorials to just learn about the tool. Having precomputed predictions makes using FiftyOne in this tutorial way easier. For example, if someone just wants to poke around Open Images. 

(^this is addressed by providing a zip)

2) It doesn't let you use your model directly anyway. The provided code lets you generate predictions for any `tensorflow_hub` model, but most people don't use that and would have to generate predictions off of the `csv` file structure you presented. It would make it a lot easier to see how the rest of the code works with example predictions to let people know if they want to go through the trouble of generating their own.

(^not addressed by zip)

3) Cutting down on bash code that needs to be run would make it more feasible to turn this into a python notebook where you copy the bash blocks. This would let the notebook require less setup.

(^not addressed by zip)

tylerganter replied:
if you trust me...let's get this merged as is, and once we've sorted out hosting I'll put a predictions file up on the website for download (and update this tutorial)

ehofesmann commented on docs/source/tutorials/open_images_evaluation.ipynb:
````diff
@@ -0,0 +1,444 @@
+    "Open up a `python` or `ipython` terminal and run the following:\n",
+    "\n",
+    "```python\n",
+    "import fiftyone as fo\n",
````
I don't see any reason why this shouldn't be a cell I can run.

tylerganter replied:
Because you are using a bash kernel, not a python kernel

ehofesmann replied:
Right, that makes a lot of sense.... This makes it feel really weird though, having a python notebook that can only run bash code blocks and you need to copy and paste the python code into a terminal. I feel like this tutorial could be streamlined a lot by using a python kernel, still being able to run the installs through `!pip install ...` and then just having to copy the 3 scripts (assuming you don't have to run inference anymore)

tylerganter replied:
okay I switched it to `%%bash` blocks

ehofesmann commented on docs/source/tutorials/open_images_evaluation.ipynb:
```diff
@@ -0,0 +1,444 @@
+   "source": [
+    "## Prepare the ground-truth for evaluation\n",
+    "\n",
+    "Open Images requires \"expanding the hierarchy\" if the ground-truth labels, for\n",
```
Missing something

tylerganter replied:
?

ehofesmann replied:
I see what it was now. "if the ground-truth labels" -> "of the ground-truth labels"

ehofesmann commented on docs/source/tutorials/open_images_evaluation.ipynb:
````diff
@@ -0,0 +1,444 @@
+    "`ipython` terminal and run the following:\n",
+    "\n",
+    "```python\n",
+    "import fiftyone as fo\n",
````
Should also be an executable cell.

ehofesmann commented on docs/source/tutorials/open_images_evaluation.ipynb:
```diff
@@ -0,0 +1,444 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### (optional) Visualize the data\n",
```
Add a screenshot of what this would look like.

ehofesmann replied:
Oh my god it's a gif! :1st_place_medal:

tylerganter commented:
> The gif in the tutorial card is a great touch! This probably the most involved tutorial yet, nice work!
> 
> The most crucial change is the ability to download some predictions. Tensorflow has so many versioning issues that it can be incredibly difficult do generate your own predictions. I remember we briefly discussed this earlier but I think it's basically necessary to upload a `PREDICTIONS.csv` file that you have used previously.
> 
> The second major point is adding a conclusion to this, after reading through it I really just wanted to see you do something in FiftyOne after spending all this work getting things into FiftyOne. I think just pointing out some of the false positive issues/incorrect labels from your blog post (and probably linking to your blog post) would be good.

I changed the conclusion. Added another picture and a few more teaser views

ehofesmann commented:
Weird formatting issue here: 

![image](https://user-images.githubusercontent.com/21222883/90181256-2683a180-dd7e-11ea-8910-932094152fa1.png)

tylerganter commented:
> Weird formatting issue here:
> 
> ![image](https://user-images.githubusercontent.com/21222883/90181256-2683a180-dd7e-11ea-8910-932094152fa1.png)

Are you in the notebook or did you build the docs?

ehofesmann commented:
> > Weird formatting issue here:
> > ![image](https://user-images.githubusercontent.com/21222883/90181256-2683a180-dd7e-11ea-8910-932094152fa1.png)
> 
> Are you in the notebook or did you build the docs?

in the notebook
